### PR TITLE
Better peer management

### PR DESCRIPTION
**Pull request by mfranciszkiewicz**
- `remove_on_disconnect` replaced with a removed == stored session check 
- problematic `redundant_peers` method removed
- peer management methods using a `Lock`

### CODE REVIEW

File: golem/network/p2p/p2pservice.py [hunk 1]
```diff
@@ -268,27 +272,28 @@ def remove_peer(self, peer_session):
...
-        self.suggested_conn_reverse.pop(p, None)
+        peer_id = peer_session.key_id
+        stored_session = self.peers.get(peer_id)
 
-        if p in self.free_peers:
-            self.free_peers.remove(p)
-        if p in self.peer_order:
-            self.peer_order.remove(p)
-
-        self.__send_degree()
```
Comment (badb) [inline golem/network/p2p/p2pservice.py:283]: Why do you remove this?

Comment (mfranciszkiewicz) [reply golem/network/p2p/p2pservice.py:283]: I moved it to `remove_peer_by_id` which will be called if `stored_session == peer_session`

**Comment by badb** [comment]
I got several errors on Windows in `test_peersession` `test_react_to_hello`.
Line 80-81 `key_id` is removed from `peers`.
Line 88 Actual call is 'Too many peers' not 'Duplicate peers'.

**Comment by mfranciszkiewicz** [comment]
`opt_peer_num` was missing in the mocked config object. Surprisingly it didn't matter much on Linux.

**Pull request merged**